### PR TITLE
[CLC-219]: Add Check Updates

### PR DESCRIPTION
**Pull request by kutluhanmetin**

### CODE REVIEW

File: base/commands/shell.go [hunk 1]
```diff
@@ -33,6 +33,11 @@ const banner = `Hazelcast CLC %s (c) 2023 Hazelcast Inc.
 
 `
 
+const newVersion = `A newer version of CLC is available!
+Click to read release notes and download: https://github.com/hazelcast/hazelcast-commandline-client/releases/%s
```
Comment (yuce) [inline base/commands/shell.go:37]:
```suggestion
const newVersionWarning = `
A newer version of CLC is available.

Visit the following link for release notes and to download:
https://github.com/hazelcast/hazelcast-commandline-client/releases/%s
```

File: Makefile [hunk 2]
```diff
@@ -1,7 +1,7 @@
 .PHONY: build test test-cover view-cover
 
 GIT_COMMIT = $(shell git rev-parse HEAD 2> /dev/null || echo unknown)
-CLC_VERSION ?= v0.0.0-CUSTOMBUILD
+CLC_VERSION ?= 0.0.0
```
Comment (yuce) [inline Makefile:4]: Is this change necessary?

File: base/commands/shell.go [hunk 3]
```diff
@@ -80,6 +85,7 @@ func (cm *ShellCommand) ExecInteractive(ctx context.Context, ec plug.ExecContext
 			logLevel := strings.ToUpper(ec.Props().GetString(clc.PropertyLogLevel))
 			logText = fmt.Sprintf("Log %9s : %s", logLevel, logPath)
 		}
+		maybePrintNewerVersion(ec)
```
Comment (yuce) [inline base/commands/shell.go:88]: Could you move this line after the following, so the update notification is displayed after the banner?

File: internal/github.go [hunk 4]
```diff
@@ -0,0 +1,31 @@
+package internal
+
+import (
+	"encoding/json"
+	"errors"
+	"io"
+	"net/http"
+)
+
+const latestRelease = "https://api.github.com/repos/hazelcast/hazelcast-commandline-client/releases/latest"
```
Comment (yuce) [inline internal/github.go:10]: How about `latestReleaseURL`?

**Review by yuce — CHANGES_REQUESTED** [review]
Some ideas...

File: base/commands/shell.go [hunk 5]
```diff
@@ -212,6 +218,17 @@ func convertStatement(stmt string) (string, error) {
 	return stmt, nil
 }
 
+func maybePrintNewerVersion(ec plug.ExecContext) {
```
[the remark this answers is not in the document]
Comment (kutluhanmetin) [reply base/commands/shell.go:224]: I cannot add it as a boolean variable:
```
 -X definition
        add string value definition of the form importpath.name=value
```

File: base/commands/update_check.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+	"time"
+
+	"github.com/hazelcast/hazelcast-commandline-client/clc/paths"
+	"github.com/hazelcast/hazelcast-commandline-client/clc/store"
+	"github.com/hazelcast/hazelcast-commandline-client/internal"
+	. "github.com/hazelcast/hazelcast-commandline-client/internal/check"
+	"github.com/hazelcast/hazelcast-commandline-client/internal/plug"
+)
+
+const skipUpdateCheck = "CLC_SKIP_UPDATE_CHECK"
```
Comment (yuce) [inline base/commands/update_check.go:18]: Could you add this environment variable to the environment variable reference in the docs?

File: Makefile [hunk 7]
```diff
@@ -2,6 +2,8 @@
 
 GIT_COMMIT = $(shell git rev-parse HEAD 2> /dev/null || echo unknown)
 CLC_VERSION ?= v0.0.0-CUSTOMBUILD
+IS_CHECK_VERSION ?= enabled
```
Comment (yuce) [inline Makefile:5]:
```suggestion
CLC_SKIP_UPDATE_CHECK ?= 0
```

File: Makefile [hunk 8]
```diff
@@ -2,6 +2,8 @@
...
+LDFLAGS = "-s -w -X 'github.com/hazelcast/hazelcast-commandline-client/internal.GitCommit=$(GIT_COMMIT)' -X 'github.com/hazelcast/hazelcast-commandline-client/internal.Version=$(CLC_VERSION)' -X 'github.com/hazelcast/hazelcast-go-client/internal.ClientType=CLC' -X 'github.com/hazelcast/hazelcast-go-client/internal.ClientVersion=$(CLC_VERSION)' -X 'github.com/hazelcast/hazelcast-commandline-client/internal.IsCheckVersion=$(ENABLE_CHECK_VERSION)'"
```
Comment (yuce) [inline Makefile:6]:
```suggestion
LDFLAGS = "-s -w -X 'github.com/hazelcast/hazelcast-commandline-client/internal.GitCommit=$(GIT_COMMIT)' -X 'github.com/hazelcast/hazelcast-commandline-client/internal.Version=$(CLC_VERSION)' -X 'github.com/hazelcast/hazelcast-go-client/internal.ClientType=CLC' -X 'github.com/hazelcast/hazelcast-go-client/internal.ClientVersion=$(CLC_VERSION)' -X 'github.com/hazelcast/hazelcast-commandline-client/internal.SkipUpdateCheck=$(CLC_SKIP_UPDATE_CHECK)'"
```

File: base/commands/update_check.go [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+	}
+	if isSkip {
+		return nil
+	}
+	v, err := internal.LatestReleaseVersion()
+	if err != nil {
+		return err
+	}
+	if v != "" && internal.CheckVersion(trimVersion(v), ">", trimVersion(internal.Version)) {
+		I2(fmt.Fprintf(ec.Stdout(), newVersionWarning, v))
```
Comment (yuce) [inline base/commands/update_check.go:46]: Let's use `ec.PrintlnUnnecessary`.

File: base/commands/update_check.go [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+	}
+	if err = updateNextCheckTime(sa); err != nil {
+		return err
+	}
+	return nil
+}
+
+func isSkipNewerVersion(sa *store.StoreAccessor) (bool, error) {
```
Comment (yuce) [inline base/commands/update_check.go:54]: How about `shouldSkipNewerVersion` ?

File: base/commands/update_check.go [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+	if internal.Version == internal.UnknownVersion {
+		return true, nil
+	}
+	if strings.Contains(internal.Version, internal.CustomBuildSuffix) {
+		return true, nil
+	}
+	if internal.IsCheckVersion == "disabled" {
+		return true, nil
+	}
```
Comment (yuce) [inline base/commands/update_check.go:63]:
```suggestion
	if internal.SkipVersionCheck == "1" {
		return true, nil
	}
```

**Review by yuce — CHANGES_REQUESTED** [review]
A few ideas, will continue to review...

File: base/commands/shell.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (yuce) [reply base/commands/shell.go]: That's right, `internal. EnableCheckVersion` should be a string, so when its value is `false`, version check should be skipped.

[the remark this answers is not in the document]
Comment (yuce) [reply base/commands/shell.go]: In the rest of the comments we decided to use `internal.SkipUpdateCheck` instead of `internal.EnableCheckVersion` and use `0` and `1` for consistency with the environment variable.

File: base/commands/update_check.go [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+
+const skipUpdateCheck = "CLC_SKIP_UPDATE_CHECK"
+
+const newVersionWarning = `
+A newer version of CLC is available.
+
+Visit the following link for release notes and to download:
+https://github.com/hazelcast/hazelcast-commandline-client/releases/%s
+
+`
```
Comment (yuce) [inline base/commands/update_check.go:26]:
```suggestion
const newVersionWarning = `A newer version of CLC is available.
Visit the following link for release notes and to download:
https://github.com/hazelcast/hazelcast-commandline-client/releases/%s
`
```

File: base/commands/update_check.go [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+
+const (
+	updateCheckKey   = "update.nextCheckTime"
+	updateVersionKey = "update.latestVersion"
+	checkInterval    = time.Hour * 24 * 7
+)
+
+func maybePrintNewerVersion(ec plug.ExecContext) error {
+	sa := store.NewStoreAccessor(filepath.Join(paths.Caches(), "version_update"), ec.Logger())
```
Comment (yuce) [inline base/commands/update_check.go:35]:
```suggestion
	sa := store.NewStoreAccessor(filepath.Join(paths.Caches(), "update"), ec.Logger())
```

File: base/commands/update_check.go [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+https://github.com/hazelcast/hazelcast-commandline-client/releases/%s
+`
+
+const (
+	updateCheckKey   = "update.nextCheckTime"
+	updateVersionKey = "update.latestVersion"
+	checkInterval    = time.Hour * 24 * 7
+)
+
+func maybePrintNewerVersion(ctx context.Context, ec plug.ExecContext) error {
```
Comment (yuce) [inline base/commands/update_check.go:32]:
```suggestion
func maybePrintNewVersionNotification(ctx context.Context, ec plug.ExecContext) error {
```
Could you add a test for this function?

Comment (kutluhanmetin) [reply base/commands/update_check.go]: Added a test case that does not make a call to the Github API: https://github.com/hazelcast/hazelcast-commandline-client/pull/294/commits/23466d7ad9ddc302f3576b5e39f1690f0be76ef5.

**Review by yuce — APPROVED** [review]
Looks good!
